src/include/storage/index/hash_index.h

```diff
@@ -0,0 +1,272 @@
+    }
+  };
+
+  using ValueMap = std::unordered_set<TupleSlot, TupleSlotHash>;
+  using ValueType = std::variant<TupleSlot, ValueMap>;
+
+  HashIndex(const catalog::index_oid_t oid, const ConstraintType constraint_type, IndexMetadata metadata)
+      : Index(constraint_type, std::move(metadata)),
+        hash_map_{new cuckoohash_map<KeyType, ValueType>(INITIAL_CUCKOOHASH_MAP_SIZE)} {}
+
+  cuckoohash_map<KeyType, ValueType> *const hash_map_;
+
+ public:
+  ~HashIndex() final { delete hash_map_; }
+
+  void PerformGarbageCollection() final{};
+
+  bool Insert(transaction::TransactionContext *const txn, const ProjectedRow &tuple, const TupleSlot location) final {
+    TERRIER_ASSERT(GetConstraintType() == ConstraintType::DEFAULT,
+                   "This Insert is designed for secondary indexes with no uniqueness constraints.");
+    KeyType index_key;
+    index_key.SetFromProjectedRow(tuple, metadata_);
+
+    bool UNUSED_ATTRIBUTE insert_result = false;
+
+    auto key_found_fn = [location, &insert_result](ValueType &value) -> bool {
+      if (std::holds_alternative<TupleSlot>(value)) {
+        // replace it with a ValueMap
+        const auto existing_location = std::get<TupleSlot>(value);
+        value = ValueMap({{location}, {existing_location}}, 2);
+        insert_result = true;
+      } else {
+        // insert the location to the cuckoohash_map
+        auto &value_map = std::get<ValueMap>(value);
+        insert_result = value_map.emplace(location).second;
+      }
+      return false;
+    };
+
+    const bool UNUSED_ATTRIBUTE uprase_result = hash_map_->uprase_fn(index_key, key_found_fn, location);
+
+    TERRIER_ASSERT(insert_result != uprase_result,
+                   "Either a new key was inserted (uprase_result), or the value already existed and a new value was "
+                   "inserted (insert_result).");
+
+    // Register an abort action with the txn context in case of rollback
+    txn->RegisterAbortAction([=]() {
+      auto abort_key_found_fn = [location](ValueType &value) -> bool {
+        if (std::holds_alternative<TupleSlot>(value)) {
+          // It's just a TupleSlot, functor should return true for uprase to erase it
+          return true;
+        }
+        auto &value_map = std::get<ValueMap>(value);
+        if (value_map.size() == 1) {
+          // ValueMap only has 1 element, functor should return true for uprase to erase it
+          TERRIER_ASSERT(value_map.count(location) == 1, "location must be the only value in the ValueMap.");
+          return true;
+        }
+        // ValueMap contains multiple elements, erase the element for location
+        const auto UNUSED_ATTRIBUTE erase_result = value_map.erase(location);
+        TERRIER_ASSERT(erase_result == 1, "Erasing from the ValueMap should not fail.");
+        return false;
+      };
+
+      const bool UNUSED_ATTRIBUTE uprase_result = hash_map_->uprase_fn(index_key, abort_key_found_fn);
```

Comment:
You should have a comment before this line that (both visually and conceptually) describes how the lambda above gets passed to the library.  Without the comment it is way too easy to miss the fact that there is a lambda defined above.